### PR TITLE
refactor: [TECH-1508][DHIS2-15103] extract search component

### PR DESCRIPTION
**Pull request by simonadomnisoru**
https://dhis2.atlassian.net/browse/TECH-1508. The program stage WL cypress tests failures are [known errors](https://github.com/dhis2/capture-app/wiki/Pull-Request-Guidelines#cypress-tests-to-follow-up-on-the-tests-are-commented-out-from-our-code-base) and will be fixed in https://github.com/dhis2/capture-app/pull/3269.

This refactor also fixes the [DHIS2-15103](https://dhis2.atlassian.net/browse/DHIS2-15103) bug

[DHIS2-15103]: https://dhis2.atlassian.net/browse/DHIS2-15103?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by eirikhaugstulen — COMMENTED** [review]
Hey @simonadomnisoru,
This is a big upgrade! I've left some comments that are probably not important now, but points for further refinement in the future. Other than that, this is great!

File: src/core_modules/capture-core/components/SearchBox/SearchBox.epics.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// @flow
+import { ofType } from 'redux-observable';
+import { EMPTY } from 'rxjs';
+import { switchMap } from 'rxjs/operators';
+import { searchBoxActionTypes } from './SearchBox.actions';
+import { buildUrlQueryString } from '../../utils/routing';
+
+export const navigateToNewUserPageEpic = (action$: InputObservable, store: ReduxStore, { history }: ApiUtils) =>
+    action$.pipe(
+        ofType(searchBoxActionTypes.NAVIGATE_TO_NEW_USER_PAGE),
```
[the remark this answers is not in the document]
Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/SearchBox/SearchBox.epics.js:10]: I agree that the word 'user' is weird, but based on the [designs](https://docs.google.com/document/d/1KNn42WDEj9cM4mDpW0RaKUi1jQpQ58EwXTFu0fyXDg4/edit#heading=h.j4q0uihpwsqv), this is how it's supposed to be. I'll follow up with the product and double-check if this is the desired behavior. Thanks!

File: src/core_modules/capture-core/components/SearchBox/SearchBox.component.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                                    navigateToRegisterUser={navigateToRegisterUser}
+                                    showInitialSearchBox={showInitialSearchBox}
+                                    fallbackTriggered={fallbackTriggered}
+                                />
+                            </div>
+                        </div>
+                    </Paper>
+                    <div className={classes.quarter}>
+                        <TemplateSelector />
+                    </div>
```
[the remark this answers is not in the document]
Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/SearchBox/SearchBox.component.js:162]: That's a great suggestion. I extracted `TemplateSelector` into the components folder. Thanks!

File: src/core_modules/capture-core/components/SearchBox/hooks/index.js
[the remark this answers is not in the document]
Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/SearchBox/hooks/index.js:1]: Good catch! Updated the code

File: src/core_modules/capture-core/components/SearchBox/SearchBox.epics.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/SearchBox/SearchBox.epics.js]: @eirikhaugstulen turns out this was a bug. It should be fixed now

File: src/core_modules/capture-core/components/SearchBox/SearchBox.component.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoakimSM) [reply src/core_modules/capture-core/components/SearchBox/SearchBox.component.js:162]: I like it @eirikhaugstulen.

@simonadomnisoru: Would be great if we can also change the insertion point for the `TemplateSelector` to the MainPage and SearchPage. I.e. not render `TemplateSelector` from within the `Searchbox`. 

I had a look at the Mainpage.container/Mainpage.component: Can we try to restructure these files a bit.
- Move TopBar to MainPage.component
- Move the logic related to if we want to display workingLists or search to the MainPage.component. The `showMainPage` function could use a refactor (Is the logic still sound? Does it also apply to the Search form? Should be moved to MainPage.component and renamed).

File: src/core_modules/capture-core/components/SearchBox/SearchBox.container.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    navigateToNewTrackedEntityPage,
+    showInitialViewOnSearchBox,
+} from './SearchBox.actions';
+import { useCurrentTrackedEntityTypeId } from '../../hooks';
+import { ResultsPageSizeContext } from '../Pages/shared-contexts';
+import { usePreselectedProgram } from './hooks';
+
+export const SearchBox = () => {
+    const dispatch = useDispatch();
+    const { programId } = useLocationQuery();
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/SearchBox/SearchBox.container.js:18]: Great if we can pass in the programId as a prop

File: src/core_modules/capture-core/components/SearchBox/SearchBox.container.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        [dispatch]);
+    const dispatchNavigateToNewTrackedEntityPage = useCallback(() => {
+        dispatch(navigateToNewTrackedEntityPage());
+    }, [dispatch]);
+
+    const { searchStatus, error, ready, searchableFields, minAttributesRequiredToSearch } = useSelector(
+        ({ searchDomain, activePage }) => ({
+            searchStatus: searchDomain.searchStatus,
+            ready: !activePage.isLoading && !activePage.lockedSelectorLoads,
+            error: activePage.selectionsError && activePage.selectionsError.error,
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/SearchBox/SearchBox.container.js:36]: Can you look into the logic here. Are these needed? `activePage` is a concept we should get rid of. We should strive for the following flow: programId as input (and maybe others that I can't think of?) -> SearchBox components get the metadata they need internally based on the input. Feel free to use the Redux store for now (except `activePage`).

Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/SearchBox/SearchBox.container.js]: Code updated @JoakimSM. Thanks for the feedback!

**Review by eirikhaugstulen — APPROVED** [review]
Looks really good, @simonadomnisoru 👍

**Review by geethaalwan — APPROVED** [review]
Tested successfully on 2.41,2.40.1,2.39.3,2.38.4 versions

**Pull request merged**